### PR TITLE
Generic interfaces on examples

### PR DESCRIPTION
**Pull request by abefernan**
Part of #399.

This PR deals with the "Interfaces" part of the issue above.

It also makes the affected contracts and tests use `cosmwasm_std::Empty` for their different implementations of the now generic interfaces.

### CODE REVIEW

File: examples/contracts/cw1-subkeys/src/cw1.rs [hunk 1]
```diff
@@ -7,6 +7,9 @@ use crate::error::ContractError;
 
 impl Cw1 for Cw1SubkeysContract<'_> {
```
Comment (hashedone) [inline examples/contracts/cw1-subkeys/src/cw1.rs:8]: Just to confirm - the contract should be generic over exec and query, but I assume there would be separate PR for that, right?

File: examples/contracts/custom/src/cw1.rs [hunk 2]
```diff
@@ -1,11 +1,14 @@
...
 use cw1::{CanExecuteResp, Cw1};
 use sylvia::types::{ExecCtx, QueryCtx};
 
 use crate::contract::CustomContract;
 
 impl Cw1 for CustomContract {
     type Error = StdError;
+    type ExecC = Empty;
+    type QueryC = Empty;
+    type CosmosCustomMsg = Empty;
```
Comment (hashedone) [inline examples/contracts/custom/src/cw1.rs:11]: What is this one about? @jawoznia do we need this for anything? I see it in more places so maybe I miss something?

**Review by hashedone — APPROVED** [review]
In general looks ok, but I see couple `CosmosCustomMsg` and I don't know what are they about. Do we need those?

File: examples/contracts/cw1-subkeys/src/cw1.rs (hunk 1 above)
Comment (abefernan) [reply examples/contracts/cw1-subkeys/src/cw1.rs:8]: Yes, this PR only deals with interfaces

File: examples/contracts/custom/src/cw1.rs (hunk 2 above)
Comment (jawoznia) [reply examples/contracts/custom/src/cw1.rs:11]: `ExecC` and `QueryC` are special associated types by which you can define the `CustomMsg` and `CustomQuery` types.

The `CosmosCustomMsg` is not needed here judging by the name.

File: examples/interfaces/cw1/src/lib.rs [hunk 3]
```diff
@@ -11,35 +11,43 @@ pub struct CanExecuteResp {
...
 
     /// Execute requests the contract to re-dispatch all these messages with the
     /// contract's address as sender. Every implementation has it's own logic to
     /// determine in
     #[sv::msg(exec)]
-    fn execute(&self, ctx: ExecCtx, msgs: Vec<CosmosMsg>) -> Result<Response, Self::Error>;
+    fn execute(
+        &self,
+        ctx: ExecCtx<Self::QueryC>,
+        msgs: Vec<CosmosMsg<Self::CosmosCustomMsg>>,
```
Comment (jawoznia) [inline examples/interfaces/cw1/src/lib.rs:27]:
```suggestion
        msgs: Vec<CosmosMsg<Self::ExecC>>,
```

File: examples/interfaces/cw1/src/lib.rs [hunk 4]
```diff
@@ -11,35 +11,43 @@ pub struct CanExecuteResp {
 }
 
 #[interface]
-#[sv::custom(msg=cosmwasm_std::Empty, query=cosmwasm_std::Empty)]
 pub trait Cw1 {
     type Error: From<StdError>;
+    type ExecC: CustomMsg;
+    type QueryC: CustomQuery;
+    type CosmosCustomMsg: CustomMsg;
```
Comment (jawoznia) [inline examples/interfaces/cw1/src/lib.rs:18]:
```suggestion
```

**Review by jawoznia — COMMENTED** [review]
`ExecC` is a type meant to be used for defining the custom message in the interface.

File: examples/contracts/custom/src/cw1.rs (hunk 2 above)
Comment (abefernan) [reply examples/contracts/custom/src/cw1.rs:11]: As discussed in Slack, it seems needed for now since it wouldn't compile.

**Comment by abefernan** [comment]
> In general looks ok, but I see couple `CosmosCustomMsg` and I don't know what are they about. Do we need those?

It seems like it's needed for now or else it won't compile. @jawoznia and @kulikthebird think it might be a bug in Sylvia

File: examples/interfaces/cw1/src/lib.rs (hunk 3 above)
Comment (jawoznia) [inline examples/interfaces/cw1/src/lib.rs:27]: Didn't we want to make it

```suggestion
        msgs: Vec<CosmosMsg<Self::ExecC>>,
```

File: examples/interfaces/cw1/src/lib.rs [hunk 5]
```diff
@@ -11,35 +11,43 @@ pub struct CanExecuteResp {
...
     /// If CanExecute returns true then a call to `Execute` with the same message,
     /// from the given sender, before any further state changes, should also succeed.
     #[sv::msg(query)]
     fn can_execute(
         &self,
-        ctx: QueryCtx,
+        ctx: QueryCtx<Self::QueryC>,
         sender: String,
-        msg: CosmosMsg,
+        msg: CosmosMsg<Self::CosmosCustomMsg>,
```
Comment (jawoznia) [inline examples/interfaces/cw1/src/lib.rs:38]: same `ExecC`

File: examples/interfaces/cw4/src/lib.rs [hunk 6]
```diff
@@ -1,44 +1,81 @@
...
-#[sv::custom(msg=cosmwasm_std::Empty, query=cosmwasm_std::Empty)]
 pub trait Cw4 {
     type Error: From<StdError>;
+    type ExecC: CustomMsg;
+    type QueryC: CustomQuery;
+    type MemberCustomMsg: CustomMsg;
+    type ListMembersCustomMsg: CustomMsg;
+    type TotalWeightCustomMsg: CustomMsg;
+    type AdminCustomMsg: CustomMsg;
+    type HooksCustomMsg: CustomMsg;
```
Comment (jawoznia) [inline examples/interfaces/cw4/src/lib.rs:14]: Remove.

The queries most likely should not return the `Response` and instead some custom response type

Please add those messages
https://github.com/CosmWasm/cw-plus/blob/main/packages/cw4/src/query.rs

**Review by jawoznia — CHANGES_REQUESTED** [review]

File: examples/interfaces/cw1/src/lib.rs (hunk 3 above)
Comment (abefernan) [reply examples/interfaces/cw1/src/lib.rs:27]: If I change this line I get `associated type ExecC not found for Self`.

And also the `#[interface]` directive errors with:

```
type alias takes 1 generic argument but 0 generic arguments were supplied
expected 1 generic argumentrustc
lib.rs(13, 1): type alias defined here, with 1 generic parameter: `ExecC`
lib.rs(13, 2): add missing generic argument: `ExecC`
```

Do I need to use the sv::custom directive?

Comment (jawoznia) [reply examples/interfaces/cw1/src/lib.rs:27]: Ok, I thought the [fix](https://github.com/CosmWasm/sylvia/pull/406) landed in this PR.

Leave it as it is and remove `CosmosCustomMsg` in following PR then.

**Review by jawoznia — APPROVED** [review]
LGTM
